src/H5Rint.c

```diff
@@ -1209,7 +1209,7 @@ H5R__decode_region(const unsigned char *buf, size_t *nbytes, H5S_t **space_ptr)
         HGOTO_ERROR(H5E_REFERENCE, H5E_CANTDECODE, FAIL, "Buffer size is too small")
     if (H5S_set_extent_simple(space, rank, NULL, NULL) < 0)
         HGOTO_ERROR(H5E_REFERENCE, H5E_CANTSET, FAIL, "can't set extent rank for selection")
-    if (H5S_SELECT_DESERIALIZE(&space, &p) < 0)
+    if (H5S_SELECT_DESERIALIZE(&space, &p, *nbytes - 2 * sizeof(uint32_t)) < 0)
```

Comment:
I'd suggest calculating an end_ptr in this function and then making the buffer size calculation here based on that rather than manual subtraction of sizes so that any future revisions to the file format don't miss updating this calculation.